system/security/plugins/jwtSecurity/jwtCache.cpp

```diff
@@ -0,0 +1,410 @@
+        else
+            scopePermMap[rewrittenScope] = static_cast<SecAccessFlags>(existingFlags | accessFlag);
+    }
+
+    return *this;
+}
+
+JWTUserInfo& JWTUserInfo::_addScopePerm(PermissionMap& scopePermMap, const std::string& scope, const std::string& accessFlagName, SecAccessFlags defaultFlag, bool asDeny)
+{
+    return _addScopePerm(scopePermMap, scope, getSecAccessFlagValue(accessFlagName.c_str()), defaultFlag, asDeny);
+}
+
+SecAccessFlags JWTUserInfo::_matchScopePerm(const PermissionMap& scopePermMap, const std::string& scope, SecAccessFlags defaultFlag) const
+{
+    // Ideally, this function should be traversing a tree of scope permissions, with each
+    // scope level a tree node; that would be much faster and would also avoid requiring
+    // ordering multiple patterns prior to calling _addScopePerm()
+    std::string         scopeAsExample(_convertPathname(_globToExample(scope)));
+    SecAccessFlags      accessValue = defaultFlag;
+    int                 longestScopeLen = 0;
+
+    for (PermissionMap::const_iterator x = scopePermMap.begin(); x != scopePermMap.end(); x++)
+    {
+        if (::fnmatch(x->first.c_str(), scopeAsExample.c_str(), (FNM_CASEFOLD | FNM_LEADING_DIR)) == 0)
+        {
+            // Use only the longest (most precise) scope
+            if (x->first.size() > longestScopeLen)
+            {
+                longestScopeLen = x->first.size();
+                accessValue = x->second;
+            }
+        }
+    }
+
+    return accessValue;
+}
+
+JWTUserInfo& JWTUserInfo::_eraseScopePerm(PermissionMap& scopePermMap, const std::string& scope)
+{
+    std::string         rewrittenScope(_convertPathname(scope));
+
+    scopePermMap.erase(rewrittenScope);
+
+    return *this;
+}
+
+JWTUserInfo& JWTUserInfo::_clearScopePerm(PermissionMap& scopePermMap)
+{
+    scopePermMap.clear();
+
+    return *this;
+}
+
+unsigned int JWTUserInfo::_countScopePerms(const PermissionMap& scopePermMap) const
+{
+    return scopePermMap.size();
+}
+
+//=============================================================================
+
+JWTUserCache::JWTUserCache()
+{
+}
+
+JWTUserCache::~JWTUserCache()
+{
+}
+
+bool JWTUserCache::has(const std::string& userName) const
+{
+    bool wasFound = false;
+
+    {
+        CriticalBlock block(crit);
+
+        wasFound = userPermMap.find(userName) != userPermMap.end();
+    }
+
+    return wasFound;
+}
+
+JWTUserCache& JWTUserCache::set(const std::string& userName, const JWTUserInfo& userInfo)
+{
+    {
+        CriticalBlock block(crit);
+
+        userPermMap[userName] = userInfo;
+    }
+
+    return *this;
+}
+
+bool JWTUserCache::get(const std::string& userName, JWTUserInfo& userInfo)
+{
+    UserPermissionMap::iterator foundIter;
+
+    {
+        CriticalBlock block(crit);
+
+        foundIter = userPermMap.find(userName);
+    }
+
+    if (foundIter != userPermMap.end())
+    {
+        userInfo = foundIter->second;
+        return true;
+    }
```

Comment:
This looks to be unsafe. Once the guard around find is released, the contents of what was found can be changing during the copy. For example, before foundIter->second is copied foundIter could be erased.

Reply:
Good catch.  Fixed.